src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -1721,7 +1721,7 @@ trait Types
           val varToParam = new TypeMap {
             def apply(tp: Type) = tp match {
               case tv: TypeVar => // Applying a type constructor variable to arguments results in a new instance of AppliedTypeVar each time
-                val toOrigin = appliedType(tv.origin.typeSymbol.typeConstructor, tv.typeArgs.mapConserve(this))
+                val toOrigin = appliedType(tv.origin.typeConstructor, tv.typeArgs.mapConserve(this))
```

Comment:
Tricky stuff! This also hints at a deeper problem with this `varToParam/paramToVar` approach: it would be nice if the map were invertible (`paramToVar(varToParam(tp)) == tp`), but that's not possible if we collapse multiple `TypeVar`s with the same origin (e.g. `WildcardType`). Since we're really just trying to avoid having unexpected `TypeVar`s in our BTS (a fundamental operation!), maybe something like the following?

```
// After this TypeMap, it's safe to recurse (`tpe.parents exists typeContainsTypeVar` above is `false`)
val varToParam = new TypeMap {
  def apply(tp: Type) = tp match {
    case tv: TypeVar => 
      tv.origin match {
        // Usually a TypeVar is derived from a type param -- map back to it, so that `paramToVar` has a unique handle to recover the TypeVar 
        case TypeRef(pre, sym, _) if sym.isTypeParameterOrSkolem =>
          // Applying a type constructor variable to arguments results in a new instance of AppliedTypeVar each time
          val toOrigin = 
            if (tv.typeArgs.isEmpty) tv.origin 
            else TypeRef(pre, sym, tv.typeArgs.mapConserve(this))
          tvarFor(toOrigin) = tv
          toOrigin
        case _ =>
          // This case is unlikely, but it happens: see e.g. wildcardToTypeVarMap.
          // We can't record in `tvarFor` map, because the TypeVar was not derived from a type param
          // and thus we're not sure its origin uniquely identifies this TypeVar.
          // Since our main goal is to eliminate `TypeVar`s, rewrite to origin and :pray:.
          tv.origin
      }
    case _ => tp.mapOver(this)
  }
}
``` 

I'll do some more digging in the history to see what we used to do, but first lunch.

Reply:
Looks like this was introduced to fix https://github.com/scala/bug/issues/2311.

I wonder if we can use something like `suspendingTypeVars` as an alternative for the original fix? I haven't looked at what the root cause of that old bug is. See also https://github.com/scala/scala/pull/7978 for a recent rework of `suspendingTypeVars`.

Perhaps a full fix is 2.14 or even 3.0 territory, and in the mean time just not crashing is enough :-)

Reply:
Thanks for the explanations. I need some more time to try and figure out how the different parts interact.

Reply:
In this case the wildcards come from implicit search:
`val wildPt = approximate(pt)`

Reply:
Actually I have no clue what is the purpose of wildcardToTypeVarMap. It's called only once and removing it seems to have no effect. It's hard to track the history of it because there have been large refactorings in the past.